added per tier stats

guptask opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/cachelib/51)
<!-- Reviewable:end -->

igchor committed:
Initial multi-tier support implementation

igchor committed:
Adjust and enable tests for ShmFileSegment

igchor committed:
Introduce FileShmSegment for file-backed shared memory

It's implementation is mostly based on PosixShmSegment.

Also, extend ShmManager and ShmSegmentOpts to support this new
segment type.

guptask committed:
Add support for shm opts serialization

After introducing file segment type, nameToKey_ does not provide
enough information to recover/remove segments on restart.

This commit fixes that by replacing nameToKey_ with nameToOpts_.

Previously, the Key from nameToKey_ map was only used in a single
DCHECK().

victoria-mcgrath committed:
Initial version of config API extension to support multiple memory tiers

* New class MemoryTierCacheConfig allows to configure a memory tier.
  Setting tier size and location of a file for file-backed memory are
  supported in this initial implementation;
* New member, vector of memory tiers, is added to class CacheAllocatorConfig.
* New test suite, chelib/allocator/tests/MemoryTiersTest.cpp,
  demonstrates the usage of and tests extended config API.

igchor committed:
Integrate Memory Tier config API with CacheAllocator.

igchor committed:
Run centos and debian workflows on push and PR

igchor committed:
Run tests on CI

igchor committed:
Add MemoryTierCacheConfig::fromShm()

to allow using new configureMemoryTiers() API with legacy behavior.

Move validation code for memory tiers to validate() method and convert
ratios to sizes lazily (on get)..

guptask committed:
Moved common segment code for posix and file shm segments into ShmCommon

igchor committed:
Fix test_shm_manager.cpp test

It wrongly assumed that the only possible segment type is
PosixSysV segment.

igchor committed:
Run long tests (navy/bench) every day on CI

victoria-mcgrath committed:
Enabled memory tier config API for cachebench.

victoria-mcgrath committed:
Enabled shared memory tier in cachebench.

victoria-mcgrath committed:
Converted nvmCacheState_ to std::optional to simplify NVM cache state handling when NVM cache state is not enabled

Commit:
Adding example for multitiered cache

igchor committed:
Extend CompressedPtr to work with multiple tiers

Now it's size is 8 bytes intead of 4.

Original CompressedPtr stored only some offset with a memory Allocator.
For multi-tier implementation, this is not enough. We must also store
tierId and when uncompressing, select a proper allocator.

An alternative could be to just resign from CompressedPtr but they
are leveraged to allow the cache to be mapped to different addresses on shared memory.

Changing CompressedPtr impacted CacheItem size - it increased from 32 to 44 bytes.

igchor committed:
Run CI on prebuild docker image

igchor committed:
Run only centos build on CI

Commit:
Implemented async Item movement between tiers

igchor committed:
Enable workarounds in tests

igchor committed:
Add basic multi-tier test

igchor committed:
Set correct size for each memory tier

igchor committed:
Extend cachbench with value validation

Commit:
Aadding new configs to hit_ratio/graph_cache_leader_fobj

igchor committed:
Move validateValue call to make sure it is measured by latency tracker

Commit:
Fix eviction flow and removeCb calls

Without this fix removeCb called even in case when Item is moved between
tiers.

igchor committed:
Remove failing build-cachelib workflow (#42)

It fails because CentOS is EOL. We might want to consider
using CentOS Streams but for now, just remove it.

Right now, we rely on build-cachelib-centos workflow anyway.

victoria-mcgrath committed:
Disabled test suite allocator-test-AllocatorTypeTest (#41)

Disabled test suite allocator-test-AllocatorTypeTest to skip sporadically failing tests.

igchor committed:
Do not compensate for rounding error when calculating tier sizes (#43)

Compensation results in ratios being different than originially specified.

victoria-mcgrath committed:
Fixed total cache size in CacheMemoryStats (#38)

Return a sum of sizes of each tier instead of just 1st tier's size.

igchor committed:
Fix tests and benchmarks compilation

Compilation of some of the tests was failing with:
"undefined reference to `facebook::cachelib::test_util::getRandomAsciiStr[abi:cxx11](unsigned int)'"

Fix this by linking those tests with common/TestUtils.cpp

Also, for some reason specyfing sources as in
add_library(common_test_support INTERFACE [sources...]) did not work.
Using target_sources works fine.

igchor committed:
Update docker file used in CI

Centos8 is EOL

igchor committed:
Disable failing clang-format-check

igchor committed:
Add one more navy test to BLACKLIST

Commit:
Merge pull request #47 from igchor/update_docker2

Update docker file used in CI

Commit:
Fix issue with "Destorying an unresolved handle"

The issue happened when ReadHandleImpl ctor needs to destroy
waitContext_ because addWaitContextForMovingItem() returns false.
So before destroying waitContext_ we are calling discard method to
notify ~ItemWaitContext() that Item is ready.

Commit:
Merge pull request #50 from vinser52/fix_unresolved_handle

Fix issue with "Destorying an unresolved handle"

guptask committed:
added per-tier stats

guptask committed:
added per-tier stats

guptask committed:
added per-tier stats

guptask committed:
added per-tier stats

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 7 files reviewed, 3 unresolved discussions (waiting on @guptask)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 3594 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo-uGg3E_nEgg9y7zr:-Mxo-uGg3E_nEgg9y7zs:b-3i6l42) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheAllocator-inl.h#L3594)):*
> ```objc
>   // initialize tier stats
>   for (auto i = 0; i < numTiers_; i++) {
>     stats_.shmTierStats.emplace_back();
> ```

I think this loop should be inside `stats_.init()` method.



_Code quote:_
```objc
  // initialize tier stats
  for (auto i = 0; i < numTiers_; i++) {
    stats_.shmTierStats.emplace_back();
  }
```
___
*[cachelib/allocator/CacheStats.h, line 316 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo07Qp86na5tWHfV_a:-Mxo07Qp86na5tWHfV_b:bhielao) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStats.h#L316)):*
> ```objc
> 
>   CacheTierStats(uint64_t evicAttempts, uint64_t evicSuccesses) :
>     numEvictionAttempts(evicAttempts), numEvictionSuccesses(evicSuccesses) {};
> ```

Why do not use default initialization of corresponding fields instead of this constructor?
___
*[cachelib/allocator/CacheStatsInternal.h, line 38 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo11WlCISG8-Ilwfd2:-Mxo11WlCISG8-Ilwfd3:b-cvdw7) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStatsInternal.h#L38)):*
> ```objc
> 
>   // number of eviction successes per tier
>   AtomicCounter numEvictionSuccesses{0};
> ```

As we discussed before on the meeting, it would be useful to see the stat of per-tier cache hits. E.g. how often we hit to a particular tier.


<!-- Sent from Reviewable.io -->

guptask reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 3594 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo-uGg3E_nEgg9y7zr:-My6ah_x6sf7JV5wcPD2:b-t7ywre) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheAllocator-inl.h#L3594)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

I think this loop should be inside `stats_.init()` method.
</blockquote></details>

fixed in latest commit
___
*[cachelib/allocator/CacheStats.h, line 316 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo07Qp86na5tWHfV_a:-My6ami2Dslnmbe0wWWv:b-a3549c) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStats.h#L316)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Why do not use default initialization of corresponding fields instead of this constructor?
</blockquote></details>

This constructor is needed for Stats::populateGlobalCacheStats(GlobalCacheStats& ret)
___
*[cachelib/allocator/CacheStatsInternal.h, line 38 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo11WlCISG8-Ilwfd2:-My6pWH99QLtR0WDrTFQ:b46t7ai) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStatsInternal.h#L38)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

As we discussed before on the meeting, it would be useful to see the stat of per-tier cache hits. E.g. how often we hit to a particular tier.
</blockquote></details>

I need to discuss where to increment this counter. I have added this and used size counters to the stats structures.


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 9 files reviewed, 7 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/Cache.h, line 279 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l7FW7DuiGcVnybOz:-My7l7FW7DuiGcVnybP-:b649m4o) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.h#L279)):*
> ```objc
>   unsigned int getNumTiers() const;
> 
>   unsigned int numTiers_ = 1;
> ```

what is the reason for adding `int`?
___
*[cachelib/allocator/Cache.cpp, line 28 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l19kB3xOa00UCsTK:-My7l19kB3xOa00UCsTL:b649m4o) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.cpp#L28)):*
> ```cpp
> CacheBase::CacheBase(unsigned numTiers): numTiers_(numTiers) {}
> 
> unsigned int CacheBase::getNumTiers() const {
> ```

what is the reason for adding `int`?
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2921 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7lDzB7k52P0zbkIpp:-My7lDzC3VX3oEu3YIEh:b-5936z3) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheAllocator-inl.h#L2921)):*
> ```objc
>   auto startTime = util::getCurrentTimeSec();
>   while (true) {
>     stats_.numEvictionAttempts.inc();
> ```

Do we need global counter on pair with per-tier?
___
*[cachelib/allocator/CacheStats.cpp, line 48 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7pV6G6i9otrH93-0f:-My7pV6G6i9otrH93-0g:b-bjeb2s) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheStats.cpp#L48)):*
> ```cpp
>   // initialize tier stats
>   for (auto i = 0; i < numTiers; i++) {
>     shmTierStats.emplace_back();
> ```

I just noticed that this loop is equivalent of the `resize()` method, isn't it?
___
*[cachelib/allocator/CacheStatsInternal.h, line 38 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo11WlCISG8-Ilwfd2:-My7p9Ee79BpkkN-Dva4:b-rgrljp) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStatsInternal.h#L38)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

I need to discuss where to increment this counter. I have added this and used size counters to the stats structures.
</blockquote></details>

Possibly it should be a thread-local counter instead of an Atomic one. Need to investigate performance impact


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 9 files reviewed, 8 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/CacheAllocator.h, line 184 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7wrWi6pH2mxTJb_IY:-My7wrWj3hTk3VbHvj1O:b9dg5ek) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheAllocator.h#L184)):*
> ```objc
>   using EventTracker = EventInterface<Key>;
> 
>   static constexpr size_t kMaxTiers = 10; // TODO: need to define right value
> ```

I think we do not need it anymore


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 10 files reviewed, 12 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 2046 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqBBYREZCbq34h1Hyg:-MyqBBYREZCbq34h1Hyh:bp9ztet) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheAllocator-inl.h#L2046)):*
> ```objc
>       ItemHandle ret;
>       ret.markExpired();
>       auto tid = getTierId(*ret);
> ```

`ret` handle is empty. How you can call getTierId(*ret) using an empty handle and why we should increment the `numAccess` counter?
Can we write a test to check this code branch?
___
*[cachelib/allocator/CacheStats.h, line 316 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqOLmj1dmDMUYDBzqj:-MyqOLmj1dmDMUYDBzqk:b-bohtur) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStats.h#L316)):*
> ```objc
> 
>   // number of access requests
>   uint64_t numAccess;
> ```

Is it cache hits?
___
*[cachelib/allocator/CacheStats.cpp, line 58 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQGav2WQbMbEVlFWI:-MyqQGaw5ArBG5kMnjXU:b-qanlj5) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStats.cpp#L58)):*
> ```cpp
> #ifndef SKIP_SIZE_VERIFY
>   // SizeVerify<sizeof(Stats)> a = SizeVerify<16144>{};
>   // std::ignore = a;
> ```

Why it is commented?
___
*[cachelib/allocator/CacheStats.cpp, line 137 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQOfn5O-_jI4wXz3l:-MyqQOfn5O-_jI4wXz3m:b-eladtc) ([raw file](https://github.com/pmem/cachelib/blob/019b2a52a1e908a7efe876ebb46162994660cc45/cachelib/allocator/CacheStats.cpp#L137)):*
> ```cpp
>   ret.numAbortedSlabReleases = numAbortedSlabReleases.get();
> }
> ```

Please revert this empty line.
___
*[cachelib/allocator/CacheStatsInternal.h, line 35 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQk723Jqoj6wigyLs:-MyqQk723Jqoj6wigyLt:bto6hhn) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStatsInternal.h#L35)):*
> ```objc
> struct ShmTierStats {
>   // number of eviction attempts per tier
>   TLCounter numEvictionAttempts{0};
> ```

In Global cache stats it is Atomic type. Does it make sense to make it `TLCounter`? When I mentioned `TLCounter`, I was talking about `numAccess`. We need to investigate performance impact of using these types.


<!-- Sent from Reviewable.io -->

guptask reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 10 files reviewed, 11 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/Cache.h, line 279 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l7FW7DuiGcVnybOz:-MyrhT0c9xyWu_nsI5Pg:baq7nyr) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.h#L279)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

what is the reason for adding `int`?
</blockquote></details>

I did it for compliance with coding standards. I think it should be size_t instead of unsigned or unsigned int. 
___
*[cachelib/allocator/Cache.cpp, line 28 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l19kB3xOa00UCsTK:-Myri4oe9enxvxasgxUF:blek9m7) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.cpp#L28)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

what is the reason for adding `int`?
</blockquote></details>

same as previous comment
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2921 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7lDzB7k52P0zbkIpp:-MyriIoR-wdNIzCgiTFv:bezuhl2) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheAllocator-inl.h#L2921)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Do we need global counter on pair with per-tier?
</blockquote></details>

This was the existing counter - it is recorder as part of SlabReleaseStats. I don't have any strong reason to retain this - can remove if you think that is a better direction. I think it will slightly boost the performance given it is an atomic counter.
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2046 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqBBYREZCbq34h1Hyg:-MyrkgdQ8wgrB9rYLWyG:b-86upl2) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheAllocator-inl.h#L2046)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

`ret` handle is empty. How you can call getTierId(*ret) using an empty handle and why we should increment the `numAccess` counter?
Can we write a test to check this code branch?
</blockquote></details>

I forgot to take this out. I didn't hit this edge case. I can write a unit test for this. In any case, this is under UNLIKELY(handle->isExpired()) check - so I don't think a cache miss requires numAccess increment.
___
*[cachelib/allocator/CacheStats.h, line 316 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqOLmj1dmDMUYDBzqj:-MyrlXgrCeTSqJNOIVFj:b-yv6yqu) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStats.h#L316)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Is it cache hits?
</blockquote></details>

yes. I'll rename it to cacheHits to remove ambiguity.
___
*[cachelib/allocator/CacheStats.cpp, line 48 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7pV6G6i9otrH93-0f:-Myrl_EiAZSpxvDCiau2:boax6uc) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheStats.cpp#L48)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

I just noticed that this loop is equivalent of the `resize()` method, isn't it?
</blockquote></details>

yes. I will try with the resize() once I fix the unit test crash I'm currently getting. I added a copy constructor in TLCounter.
___
*[cachelib/allocator/CacheStats.cpp, line 58 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQGav2WQbMbEVlFWI:-Myrm8M3CE7nXa8ApsnI:b-1irif9) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStats.cpp#L58)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Why it is commented?
</blockquote></details>

I don't think I made this change. Also since Stats has increased in size due to the tier stats, this 16144 count won't work anymore. I guess the size would change based on number of tiers. 


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 10 files reviewed, 12 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/common/AtomicCounter.h, line 77 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-Mys4VobBoCaZtIHmUz4:-Mys4VobBoCaZtIHmUz5:b-rwuoy5) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/common/AtomicCounter.h#L77)):*
> ```objc
> 
>   TLCounter(const TLCounter& rhs)
>       : val_{rhs.val_.getSnapshot()} {}
> ```

I am not sure we have to extend TLCounter implementation.


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 10 files reviewed, 10 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 2921 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7lDzB7k52P0zbkIpp:-Mys5le80EDQ2o_3Bq78:btx217f) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheAllocator-inl.h#L2921)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

This was the existing counter - it is recorder as part of SlabReleaseStats. I don't have any strong reason to retain this - can remove if you think that is a better direction. I think it will slightly boost the performance given it is an atomic counter.
</blockquote></details>

ok, let's keep it as is to make sure we are not breaking backward compatibility.
___
*[cachelib/allocator/CacheStats.cpp, line 48 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7pV6G6i9otrH93-0f:-Mys5Kzo4w38eVZXmez6:bdv39wh) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/CacheStats.cpp#L48)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

yes. I will try with the resize() once I fix the unit test crash I'm currently getting. I added a copy constructor in TLCounter.
</blockquote></details>

Added comment below that probably we should not extend TLCounter impl
___
*[cachelib/allocator/CacheStats.cpp, line 58 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQGav2WQbMbEVlFWI:-Mys4qDW5zIfVuqS0Rc9:baf6kd1) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStats.cpp#L58)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

I don't think I made this change. Also since Stats has increased in size due to the tier stats, this 16144 count won't work anymore. I guess the size would change based on number of tiers. 
</blockquote></details>

But diff indicates that this changes is part of your PR. 
You store tier info in a vector. `sizeof(vector<...>)` is a constant and does not depend on `vector.size()`


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 0 of 10 files reviewed, 10 unresolved discussions (waiting on @guptask and @vinser52)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 2046 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqBBYREZCbq34h1Hyg:-Mys5zoSEno32B2K4K9S:b-jjf4ze) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheAllocator-inl.h#L2046)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

I forgot to take this out. I didn't hit this edge case. I can write a unit test for this. In any case, this is under UNLIKELY(handle->isExpired()) check - so I don't think a cache miss requires numAccess increment.
</blockquote></details>

I agree that we should not count hit if Item is expired.


<!-- Sent from Reviewable.io -->

igchor reviewed (changes requested):
> Reviewed 1 of 5 files at r3, 2 of 6 files at r4, 3 of 5 files at r6, 1 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 6 of 10 files reviewed, 10 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[cachelib/allocator/Cache.h, line 279 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l7FW7DuiGcVnybOz:-MyvgPpOEDs7Lg7yuWyx:bqz3fcy) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.h#L279)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

I did it for compliance with coding standards. I think it should be size_t instead of unsigned or unsigned int. 
</blockquote></details>

This is kind of unrelated to the statistics so if you want to change it, do it in separate commit and then keep the history clean (squash related commits together). 
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2959 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-MyvgpP11jI_IQ4NCg6h:-MyvgpP11jI_IQ4NCg6i:balyc9e) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheAllocator-inl.h#L2959)):*
> ```objc
> 
>       stats_.numEvictionSuccesses.inc();
>       stats_.shmTierStats[tid].numEvictionSuccesses.inc();
> ```

Don't we want to have per Pool and AllocClass stats, like we have above? Why not just copy the `(*stats_.chainedItemEvictions)` and `(*stats_.regulatItemEvictions)` to per-tier stats?
___
*[cachelib/allocator/CacheStats.h, line 317 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-Myvh90O8BPY-U_cLPIG:-Myvh90O8BPY-U_cLPIH:b-idx9bi) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheStats.h#L317)):*
> ```objc
> 
>   // size of cache tier used
>   uint64_t usedSize;
> ```

is this used anywhere?
___
*[cachelib/allocator/CacheStatsInternal.h, line 41 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-MyvfDiwBaTllv5Hyy7w:-MyvfDiwBaTllv5Hyy7x:b-pzqxhw) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheStatsInternal.h#L41)):*
> ```objc
> 
>   // number of tier access
>   AtomicCounter numHits{0};
> ```

Why not keep the same naming convention as we have right now? E.g. numCacheGets, numCacheMiss, etc.?
___
*[cachelib/allocator/CacheStatsInternal.h, line 174 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-MyvfZGI-Cqgli1f1Wpk:-MyvfZGI-Cqgli1f1Wpl:b-pq9d99) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheStatsInternal.h#L174)):*
> ```objc
> 
>   // the numbers for move and evictions in the process of slab release.
>   TLCounter numMoveAttempts{0};
> ```

Why you changed those types?


<!-- Sent from Reviewable.io -->

guptask commented:
___
*[cachelib/allocator/CacheStatsInternal.h, line 35 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQk723Jqoj6wigyLs:-Mz-YsMQ138roGcsjcuu:bb8bxr0) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStatsInternal.h#L35)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

In Global cache stats it is Atomic type. Does it make sense to make it `TLCounter`? When I mentioned `TLCounter`, I was talking about `numAccess`. We need to investigate performance impact of using these types.
</blockquote></details>

Any code changes related to this should not be part of this PR. 


<!-- Sent from Reviewable.io -->

guptask commented:
___
*[cachelib/allocator/CacheStatsInternal.h, line 174 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-MyvfZGI-Cqgli1f1Wpk:-Mz-ZHCNDsSXht-0hWTo:b-bnrurk) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheStatsInternal.h#L174)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Why you changed those types?
</blockquote></details>

reverted to original counter type


<!-- Sent from Reviewable.io -->

guptask commented:
___
*[cachelib/common/AtomicCounter.h, line 77 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-Mys4VobBoCaZtIHmUz4:-Mz-ZS-MBfurnSQwXxbs:bsnc5ay) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/common/AtomicCounter.h#L77)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

I am not sure we have to extend TLCounter implementation.
</blockquote></details>

reverted this change


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
@guptask Can you please run some benchmark to see if there is any performance impact of your changes?
Can you also pase a sample output which is generated after your changes?

> Reviewed 4 of 7 files at r9, 1 of 2 files at r10, 2 of 6 files at r11.
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51#-:-MzEpSV678hklD2bvTvo:b-1i4p4e)* status: 9 of 13 files reviewed, 10 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[cachelib/allocator/memory/SlabAllocator.h, line 327 at r11](https://reviewable.io/reviews/pmem/cachelib/51#-MzEoxsrE-nCb73b2rXN:-MzEoxsrE-nCb73b2rXO:b-fican8) ([raw file](https://github.com/pmem/cachelib/blob/44d64ff8922ebbc38f07f23f5a96eceb4ffa363f/cachelib/allocator/memory/SlabAllocator.h#L327)):*
> ```objc
> 
>   // Retrive size of a pool
>   // @return size of Pool
> ```

Could you please extend this comment or change the name? It's not clear from just the name/description if this is total size of the pool or used size. In fact, 


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 9 of 13 files reviewed, 8 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[cachelib/allocator/Cache.cpp, line 28 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l19kB3xOa00UCsTK:-MzFcTXD3QrwS9V2MI3L:bfy4wpx) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.cpp#L28)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

same as previous comment
</blockquote></details>

As mentioned above by @igchor we should keep our commit history clean. This change does not related to statistics enhancement. 


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 9 of 13 files reviewed, 8 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[cachelib/allocator/CacheStatsInternal.h, line 35 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQk723Jqoj6wigyLs:-MzFcoym7twdLZ-lVMGx:b46hhoy) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStatsInternal.h#L35)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

Any code changes related to this should not be part of this PR. 
</blockquote></details>

Sorry, I did not get what you mean. I asked whether we need to use AtomicCounter or TLCounter as a type for these fields?


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/CacheStatsInternal.h, line 35 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQk723Jqoj6wigyLs:-MzX4VUk4D45le34DUfc:b-be68be) ([raw file](https://github.com/pmem/cachelib/blob/d7c82d2b6a3ecef752415ac47e854e931155c2a5/cachelib/allocator/CacheStatsInternal.h#L35)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Sorry, I did not get what you mean. I asked whether we need to use AtomicCounter or TLCounter as a type for these fields?
</blockquote></details>

Igor asked me not to make any changes (if required after testing) for this as part of this PR. I'm running tests and will update you on any performance impact I notice. If there is any significant impact, this change would be part of separate PR. 


<!-- Sent from Reviewable.io -->

guptask committed:
Merge branch 'stats' of https://github.com/guptask/CacheLib into stats

guptask committed:
blah blah

guptask committed:
blah blah

guptask committed:
Merge branch 'stats' of https://github.com/guptask/CacheLib into stats

igchor reviewed (changes requested):
@guptask Could you also paste output for the statistics for our default hit ratio config with 4GB DRAM + 4 GB PMEM? It would be easier to verify if the numbers are correct/

> Reviewed 2 of 6 files at r11, 2 of 2 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51#-:-N-7uN4w-jTwHknTAWWs:b-t0ewli)* status: 12 of 13 files reviewed, 13 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[cachelib/allocator/CacheAllocator-inl.h, line 394 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7ujV63uLqWD4EUmpj:-N-7ujV7023v7T2CymIK:brrlnmz) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L394)):*
> ```objc
>     handle = acquire(new (memory) Item(key, size, creationTime, expiryTime));
>     if (handle) {
>       stats_.shmTierStats[tid].usedSize.set(allocator_[tid]->getPoolSize(pid));
> ```

I don't understand this line. Can you explain?
___
*[cachelib/allocator/CacheAllocator-inl.h, line 1473 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7uzRrCslZzyV10i7n:-N-7uzRrCslZzyV10i7o:b-ci81rs) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L1473)):*
> ```objc
>     if(toReleaseHandle) {
>       movedToNextTier = true;
>       stats_.shmTierStats[tid].numEvictionSuccess[pid][cid].inc();
> ```

Hm, prehaps we could also have number of eviction which resulted in item being freed/evicted to SSD as opposed to being evicted to different memory tier?

How about renaming `numEvictionAttempts` to `numMemoryEvictionAttempts` and `numEvictionSuccess` to `numMemoryEvictionSuccess` and adding `numEvictionSuccess` which would mean that `advandeIteratorAdnTryEvictRegularItem succeeded?
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2059 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7vi8N4Naw29wacpxM:-N-7vi8N4Naw29wacpxN:br1ibtj) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2059)):*
> ```objc
>     }
>     auto tid = getTierId(*handle);
>     stats_.shmTierStats[tid].numGets.inc();
> ```

Coul you move it to the findFastImpl? THere's already code for increasing numCacheGets there.
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2928 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7vwRo2_owT_x06wi5:-N-7vwRo2_owT_x06wi6:b66tox7) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2928)):*
> ```objc
>     stats_.numEvictionAttempts.inc();
>     stats_.shmTierStats[tid].numEvictionAttempts.inc();
> ```

so, you are using this shmTierStats.numEvictionAttempts counter here and inside findEviction(). Is this compatible with how the global numEvictionAttempts counter is used? From what I see, the global one is not incremented inside findEviction, Maybe we should have a separate counter for findEviction and evictForSlabRelease?
___
*[cachelib/allocator/CacheAllocator-inl.h, line 2962 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7weLJE1McTdlGFmuc:-N-7weLJE1McTdlGFmud:b-bqc1ef) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2962)):*
> ```objc
> 
>       stats_.numEvictionSuccesses.inc();
>       stats_.shmTierStats[tid].numEvictionSuccess[allocInfo.poolId][allocInfo.classId].inc();
> ```

Similar as above


<!-- Sent from Reviewable.io -->

igchor committed:
Add scripts for rebuilding/pushing docker images

Taken from: https://github.com/pmem/dev-utils-kit/commit/30794c3e1bbc9273e87da3e8f3ce7e5a2792b19e

igchor committed:
Extend CI to rebuild docker automatically

igchor committed:
Add extra param to build-package.sh

igchor committed:
Merge pull request #57 from igchor/auto_docker_build

Auto docker build

guptask commented:
*[cachelib/allocator/CacheAllocator-inl.h, line 394 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7ujV63uLqWD4EUmpj:-N-MqWqgBij80YbGKEHj:bu9k2qu) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L394)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

I don't understand this line. Can you explain?
</blockquote></details>

This is the tier used size counter. We had  a discussion about the getPoolSize() API.


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/CacheAllocator-inl.h, line 2962 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7weLJE1McTdlGFmuc:-N-OgyZYBKRqz54jqlBZ:b-896fix) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2962)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Similar as above
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/CacheStats.h, line 317 at r7](https://reviewable.io/reviews/pmem/cachelib/51#-Myvh90O8BPY-U_cLPIG:-Mz-Ygcm3BOU1HL2aG2h:b-896fix) ([raw file](https://github.com/pmem/cachelib/blob/c45671c8b0fc526eb3567a97dda2d070704f3f38/cachelib/allocator/CacheStats.h#L317)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

is this used anywhere?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/CacheStats.cpp, line 137 at r3](https://reviewable.io/reviews/pmem/cachelib/51#-MyqQOfn5O-_jI4wXz3l:-Mz-YmU-AzjbF9ZssRy0:b-896fix) ([raw file](https://github.com/pmem/cachelib/blob/019b2a52a1e908a7efe876ebb46162994660cc45/cachelib/allocator/CacheStats.cpp#L137)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Please revert this empty line.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/CacheStatsInternal.h, line 38 at r1](https://reviewable.io/reviews/pmem/cachelib/51#-Mxo11WlCISG8-Ilwfd2:-Mz-ZBiS52uD1jbCZKOf:b-896fix) ([raw file](https://github.com/pmem/cachelib/blob/52c0b3ceb1c44d0231855ff1e2cf0366f7f28e68/cachelib/allocator/CacheStatsInternal.h#L38)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Possibly it should be a thread-local counter instead of an Atomic one. Need to investigate performance impact
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

guptask commented:
*[cachelib/allocator/Cache.cpp, line 28 at r2](https://reviewable.io/reviews/pmem/cachelib/51#-My7l19kB3xOa00UCsTK:-N-OhOeZBRClWqxpfwtU:b-896fix) ([raw file](https://github.com/pmem/cachelib/blob/d923a438797c0f56f496899afcacb2b03c295c06/cachelib/allocator/Cache.cpp#L28)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

As mentioned above by @igchor we should keep our commit history clean. This change does not related to statistics enhancement. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

guptask committed:
Merge branch 'develop' of https://github.com/pmem/CacheLib into stats

guptask committed:
renamed some stats

guptask commented:
*[cachelib/allocator/CacheAllocator-inl.h, line 2059 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7vi8N4Naw29wacpxM:-N-gB3OJ2UTMVhC5qsU7:b-bgpfun) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2059)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Coul you move it to the findFastImpl? THere's already code for increasing numCacheGets there.
</blockquote></details>

I don't think I can access tid from findFastImpl 


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51)* status: 7 of 13 files reviewed, 9 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)

___
*[`cachelib/allocator/CacheAllocator-inl.h` line 394 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7ujV63uLqWD4EUmpj:-N0l3xr9DDn_x7w7btFE:b-un4s07) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L394)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

This is the tier used size counter. We had  a discussion about the getPoolSize() API.
</blockquote></details>

Can you remind me why this assignment is not done inside getStats?
___
*[`cachelib/allocator/CacheAllocator-inl.h` line 2059 at r13](https://reviewable.io/reviews/pmem/cachelib/51#-N-7vi8N4Naw29wacpxM:-N0l3ZB53rRg-d9tqkCG:b-ml4r63) ([raw file](https://github.com/pmem/cachelib/blob/075edd007ea78e4f8ba93042b365a47c738b91a9/cachelib/allocator/CacheAllocator-inl.h#L2059)):*
<details><summary><i>Previously, guptask (Sounak Gupta) wrote…</i></summary><blockquote>

I don't think I can access tid from findFastImpl 
</blockquote></details>

Why not, you get a handle inside findFastImpl, no?


<!-- Sent from Reviewable.io -->

igchor reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/pmem/cachelib/51#-:-N0l4I466Tq4lIvPVrWw:bnfp4nl)* status: 7 of 13 files reviewed, 9 unresolved discussions (waiting on @guptask, @igchor, and @vinser52)



<!-- Sent from Reviewable.io -->
